[Doc] Real-time inference documentation (#1312)
awslabs/graphstorm: docs/source/cli/model-training-inference/real-time-inference.rst

.. _real-time-inference-on-sagemaker:

========================================
Real-time Inference on Amazon SageMaker
========================================

GraphStorm CLIs for model inference on :ref:`signle machine <single-machine-training-inference>`,
:ref:`distributed clusters <distributed-cluster>`, and :ref:`Amazon SageMaker <distributed-sagemaker>`
are designed to handle large datasets, which could take minutes to hours for predicting a large number of target
nodes/edges or generating embeddings for all nodes. This is typically referred to as offline inference, where the
model processes a large batch of data at once, and the results are not needed immediately.

However, in certain use cases such as recommendation systems, social network analysis, and fraud detection, there
is a need for real-time predictions. For instance, you may only want to predict a few targets and expect to get

> results immediately, say within one second. In these scenarios, you will need a 7*24 running server to host the

Reviewer: ```suggestion
results immediately to perform an action in response to a user request. In these scenarios, you will need to deploy a server to host the
```

  file, and uses it to restore the trained model weights.
- The updated graph construciton JSON file, ``data_transform_new.json``. This JSON file is one of the outputs of
  graph construction pipeline. It contains updated information about feature transformations and feature
  dimensions. If using the :ref:`Single Machine Graph Construction <single-machine-gconstruction>` (GConstruct), the
  file is saved at the path specified by the ``--output-dir`` argument. For :ref:`Distributed Graph Construction
  <distributed-gconstruction>` (GSProcessing), the file is saved at the path specified by either ``--output-data-s3``
  or ``--output-dir`` argument.
- The updated model training configuration YAML file, ``GRAPHSTORM_RUNTIME_UPDATED_TRAINING_CONFIG.yaml``. This
  YAML file is one of the outputs of model training. It contains the updated configurations of a model by
  replacing values of configuration YAML file with values given in the training CLI arguments. If set
  ``--save-model-path`` or ``--model-artifact-s3`` configuration in model training, this updated YAML file will
  be saved to the location specified.

.. note:: 

>     Starting with v0.5, GraphStorm will save both updated JSON and YAML files into the same location as trained model

Reviewer: ```suggestion
    Starting with v0.5, GraphStorm will save both updated JSON and YAML files into the same location as the trained model
```

            --role arn:aws:iam::<account_id>:role/<your_role> \
            --region <region> \
            --restore-model-path <restore-model-path>/<epoch-XX-iter-XX> \
            --model-yaml-config-file <restore-model-path>/GRAPHSTORM_RUNTIME_UPDATED_TRAINING_CONFIG.yaml \
            --graph-json-config-file <restore-model-path>/data_transform_new.json \
            --infer-task-type <task_type> \
            --upload-tarfile-s3 s3://<a-bucket> \
            --model-name <model-name>

Arguments of the launch endpoint script include:

- **-\-image-uri** (Required): the URI of your GraphStorm real-time inference Docker image built and
  pushed in the previous :ref:`Setup  GraphStorm Real-time Inference Docker Image <build_rt_inference_docker>` step.
- **-\-region** (Required): the AWS region to deploy endpoint. This region should be **same** as the ECR
  where your Docker image is stored.
- **-\-role** (Required): the role ARN that has SageMaker execution role. Please refer to the
  `SageMaker AI document <https://docs.aws.amazon.com/sagemaker/latest/dg/realtime-endpoints-deploy-models.html#deploy-prereqs>`_
  section for details.

> - **-\-instance-type**: the instance types to be used for endpoints. (Default: ``ml.c6i.xlarge``)

Reviewer: ```suggestion
- **-\-instance-type**: the instance type to be used for endpoints. (Default: ``ml.c6i.xlarge``)
```

`Amazon Neptune Database <https://aws.amazon.com/neptune/>`_, as data source for subgraph extraction.

Once the subgraph is extracted, you will need to prepare it as the payload of different APIs for `invoke 
models for real-time inference
<https://docs.aws.amazon.com/sagemaker/latest/dg/realtime-endpoints-test-endpoints.html#realtime-endpoints-test-endpoints-api>`_.
GraphStorm defines a :ref:`specification of the payload contents <rt-request_payload-spec>` for your reference.

Below is an example payload JSON object for node classification inference.

.. code:: yaml

    {
        "version": "gs-realtime-v0.1",
        "gml_task": "node_classification",
        "graph": {
            "nodes": [
                {
                    "node_type": "author",
                    "node_id": "a4444",
                    "features": {"feat_num" : [ 0.011269339360296726, ......, ],
                                 "feat_cat" : "UK"},
                },
                {
                    "node_type": "author",
                    "node_id": "a39",
                    "features": {"feat": [-0.0032965524587780237, ......, ],

>                                 "feat_cat" : "USA"},

Author: They should. Changed.

                ......
            ],
            "edges": [
                {
                    "edge_type": [
                        "author",
                        "writing",
                        "paper"
                    ],
                    "src_node_id": "p4463",
                    "dest_node_id": "p4463",
                    "features": {"feat1": [ 1.411269339360296726, ......, ]},
                                 "feat2" : "1980s"},
                },
                ......
            ]
        },
        "targets": [
            {
                "node_type": "author",
                "node_id": "a39"
            }
        ]
    }

Invoke endpoints
****************
There are multiple ways to invoke a Sagemaker real-time inference endpoint as documented in
`SageMaker Developer Guide <https://docs.aws.amazon.com/sagemaker/latest/dg/realtime-endpoints-test-endpoints.html#realtime-endpoints-test-endpoints-api>`_.

.. note:: GraphStorm real-time inference endpoints currently only support "application/json" as content type.

Here is an example of how you can read a payload from a JSON file and use the boto3 APIs to
invoke an endpoint.

.. code-block:: python

>     import boto3
>     import json

Reviewer: pep8 lib imports before 3rd party
```suggestion
    import json
    import boto3
    
```

    # Create a SageMaker client object\n",
    sagemaker = boto3.client('sagemaker')
    # Create a SageMaker runtime client object using your IAM role ARN\n",
    runtime = boto3.client('sagemaker-runtime',

>                            aws_access_key_id='your access key string',
>                            aws_secret_access_key='your secret key string',

Reviewer: Let's remove these, it's safer to let boto3 find credentials on its own rather than having them in code
```suggestion
```

>                            region_name='asw region' # e.g., us-east-1

Reviewer: ```suggestion
                           region_name='aws region' # e.g., us-east-1
```

    endpoint_name='your endpoint name'    # e.g., GSF-Model4Realtime-Endpoint-2025-07-11-21-44-36
    # load payload from a JSON file
    with open('subg.json', 'r') as f:
         payload = json.load(f)
    content_type = 'application/json'

    # invoke endpoint
    response = runtime.invoke_endpoint(
        EndpointName=endpoint_name,
        Body=json.dumps(payload),
        ContentType=content_type,
        )
    # Decodes and prints the response body
    print(response['Body'].read().decode('utf-8'))

Endpoint invoke response
*************************

As shown in the previous invoke example, the response from a GraphStorm real-time inference endpoint will include
a JSON object in the ``Body`` field of the SageMaker API response. The details of the response syntax can be found
in the :ref:`specification of realt-time inference response <rt-response-body-spec>`.

> An example of a successful inference response:

Reviewer: Can we add a brief exlanation about the contents of "prediction" and how one could use them downstream (e.g. apply argmax)

Replace the ``123456789012`` with your own AWS account ID. For more build and push options, see 
``bash docker/build_graphstorm_image.sh --help`` and ``bash docker/push_graphstorm_image.sh --help``.

.. note::

    When comparing CPU instances to GPU instances for real-time inference, CPU instances prove more
    cost-effective while maintaining similar inference latency to GPU instances. While CPU
    instances are generally recommended for real-time inference, users are encouraged to run their own
    benchmarks and cost analysis to make the final decision between CPU and GPU instances based on their
    particular workload needs.

Deploy a SageMaker Real-time Inference endpoint
------------------------------------------------

To deploy a SageMaker real-time inference endpoint, you will need three model artifacts that were generated
during  graph construction (GConstruct/GSProcessing) and model training.

- The saved model path that contains the ``model.bin`` file. This path has the same purpose as the
  ``--restore-model-path`` used during offline inference CLIs in which GraphStorm looks for the ``model.bin``
  file, and uses it to restore the trained model weights.

> - The updated graph construciton JSON file, ``data_transform_new.json``. This JSON file is one of the outputs of

Reviewer: ```suggestion
- The updated graph construciton JSON file, ``data_transform_new.json``. This JSON file is one of the outputs of
```

                {
                    "edge_type" : [
                        "author",
                        "writing",
                        "paper"
                    ],
                    "src_node_id"   : "p4463",
                    "dest_node_id"  : "p4463",
                    "features"      : {"feat1": [ 1.411269339360296726, ......, ]},
                                       "feat2" : "1980s"},
                },
                ......
            ]
        },
        "targets" : [
            {
                "node_type" : "author",
                "node_id"   : "a39"
            }
        ]
    }

Invoke endpoints
****************
There are multiple ways to invoke a Sagemaker real-time inference endpoint as documented in
`SageMaker Developer Guide <https://docs.aws.amazon.com/sagemaker/latest/dg/realtime-endpoints-test-endpoints.html#realtime-endpoints-test-endpoints-api>`_.

.. note:: GraphStorm real-time inference endpoints currently only support "application/json" as content type.

Here is an example of how you can read a payload from a JSON file and use the boto3 APIs to
invoke an endpoint.

.. code-block:: python

    import json
    import boto3
    

>     # Create a SageMaker client object\n",

Reviewer: ```suggestion
    # Create a SageMaker client object,
```

    sagemaker = boto3.client('sagemaker')
    # Create a SageMaker runtime client object using your IAM role ARN\n",
    runtime = boto3.client('sagemaker-runtime',
                           region_name='aws region' # e.g., us-east-1
    endpoint_name='your endpoint name'    # e.g., GSF-Model4Realtime-Endpoint-2025-07-11-21-44-36
    # load payload from a JSON file
    with open('subg.json', 'r') as f:
         payload = json.load(f)
    content_type = 'application/json'

    # invoke endpoint
    response = runtime.invoke_endpoint(
        EndpointName=endpoint_name,
        Body=json.dumps(payload),
        ContentType=content_type,
        )
    # Decodes and prints the response body
    print(response['Body'].read().decode('utf-8'))

Endpoint invoke response
*************************

As shown in the previous example, the response from a GraphStorm real-time inference endpoint will include
a JSON object in the ``Body`` field of the SageMaker API response. The details of the response syntax can be found

> in the :ref:`specification of realt-time inference response <rt-response-body-spec>`.

Reviewer: ```suggestion
in the :ref:`specification of real-time inference response <rt-response-body-spec>`.
```

.. _real-time-inference-on-sagemaker:

Real-time Inference on Amazon SageMaker
----------------------------------------
GraphStorm CLIs for model inference on :ref:`signle machine <single-machine-training-inference>`,
:ref:`distributed clusters <distributed-cluster>`, and :ref:`Amazon SageMaker <distributed-sagemaker>`
are designed to tackle large dataset, which could take minutes to hours for predicting a large
number of target nodes/edges or generating embeddings for all nodes. In certain cases when you want to
predict a few targets only and hope to obtain results quickly, e.g., less than one second, you will need

> a 7*24 running server to host trained model and response to inference requests in real time.

Author: real-time updates are not necessary for online-inference.
revised to reflect the difference between offline and online inference.

> Since version 0.5, GraphStorm offers CLIs to deploy a trained model as a SageMaker real-time inference

Author: sure. replaced with `new features`.

have full ECR access to be able to pull from ECR to build the image, create an ECR repository if it
doesn't exist, and push the real-time inference image to the repository. See the `official ECR docs
<https://docs.aws.amazon.com/AmazonECR/latest/userguide/image-push-iam.html>`_ for details.

In short you can run the following:

.. code-block:: bash

    git clone https://github.com/awslabs/graphstorm.git
    cd graphstorm/

    # Build the GraphStorm real-time inference Docker image to be used on CPUs
    bash docker/build_graphstorm_image.sh --environment sagemaker-endpoint --device cpu

    # Will push an image to '123456789012.dkr.ecr.us-east-1.amazonaws.com/graphstorm:sagemaker-endpoint-cpu'
    bash docker/push_graphstorm_image.sh --environment sagemaker-endpoint --device cpu --region "us-east-1" --account "123456789012"

Replace the ``123456789012`` with your own AWS account ID. For more build and push options, see 
``bash docker/build_graphstorm_image.sh --help`` and ``bash docker/push_graphstorm_image.sh --help``.

.. note::

    CPU instances are more cost-effective and have similar inference latency as GPU instances. Therefore, it is

>     recommended to use CPU instances for real-time inference.

Author: Suggest to keep this as this is a best practice from real use cases. Revise the wording to include GPU instances on consideration.

Deploy a SageMaker Real-time Inference endpoint
................................................
To deploy a SageMaker real-time inference endpoint, you will need three model artifacts that were generated
by graph construciton and model training pipelines.

- The saved model path that contains the ``model.bin`` file. This path has the same purpose as the

>   ``--restore-model-path`` used during offline inference CLIs in which GraphStorm code looks for the ``model.bin``

Reviewer 2: ```suggestion
  ``--restore-model-path`` used during offline inference CLIs in which GraphStorm looks for the ``model.bin``
```

  file, and uses it to restore a trained model.
- The updated graph construciton JSON file, ``data_transform_new.json``. This JSON file is one of the outputs of
  graph construction pipeline. It contains the updated information about feature transformation and feature
  dimensions. If using the :ref:`Single Machine Graph Construction <single-machine-gconstruction>` CLIs, the
  file is saved at the path specified by the ``--output-dir`` argument. For :ref:`Distributed Graph Construction
  <distributed-gconstruction>` CLIs, the file is saved at the path specified by either ``--output-data-s3``
  or ``--output-dir`` argument.
- The updated model training configuration YAML file, ``GRAPHSTORM_RUNTIME_UPDATED_TRAINING_CONFIG.yaml``. This
  YAML file is one of the outputs of model training pipeline. It contains the updated configurations of a model by

>   replacing values of configuration YAML file with values given in CLIs arguments. If set

Reviewer 2: ```suggestion
  replacing values of configuration YAML file with values given in the training CLI arguments. If set
```

- **-\-role** (Required): the role ARN that has SageMaker execution role. Please refer to the
  `SageMaker AI document <https://docs.aws.amazon.com/sagemaker/latest/dg/realtime-endpoints-deploy-models.html#deploy-prereqs>`_
  section for details.
- **-\-instance-type**: the instance types to be used for endpoints. (Default: ``ml.c6i.xlarge``)
- **-\-instance-count**: the number of endpoints to be deployed. (Default: 1)
- **-\-custom-production-variant**: dictionary string that includes custom configurations of the SageMaker
  ProductionVariant. For details, please refer to `ProductionVariant Documentation
  <https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_ProductionVariant.html>`_.
- **-\-async-execution**: the mode of endpoint creation. Set ``True`` to deploy endpoint asynchronously,
  or ``False`` to wait for creation completed. (Default: ``True``)
- **-\-restore-model-path** (Required): the path where the ``model.bin`` file is saved.
- **-\-model-yaml-config-file** (Required): the path where the updated model configuration YAML file is saved.
- **-\-graph-json-config-file** (Required): the path where the updated graph construction configuration JSON file

>   is saved.

Author: sure.

> - **-\-upload-tarfile-s3** (Required): the S3 location for uploading the packed and compressed model artifacts

Author: Yes, a prefix.

  tar file.
- **-\-infer-task-type** (Required): the name of real-time inference task. Options include ``node_classification``
  and ``node_regression``.
- **-\-model-name**: the name of model. This name will be used to define names of SageMaker Model,
  EndpointConfig, and Endpoint by appending datetime to this model name. The name should follow a regular

>   expression pattern: ``^[a-zA-Z0-9]([\-a-zA-Z0-9]*[a-zA-Z0-9])$``. (Default: ``GSF-Model4Realtime``)

Author: Added the link to SM.

> This command will log out the deployed endpoint name based on the value for ``--model-name``, e.g.,

Reviewer 2: What does log out mean?

Author: It means output. Will clarify.

``GSF-Model4Realtime-Endpoint-2025-06-04-23-47-11``, to be used in the invoke step. The same endpoint name
can also be found from Amazon SageMaker AI Web console under the "Inference -> Endpoints" menu.

Invoke Real-time Inference Endpoints
.....................................
For real-time inference, you will need to extract a subgraph around the target nodes/edges from a large
graph, and use the subgraph as input of model, which is similar to how models are trained. Because time is

> critical for real-time infernce, it is recommened to use OLTP graph database, e.g., Amazon Neptune Database,

Reviewer 2: Please add a link to Amazon NDB.

Author: added.

as data source for subgraph extraction. 

Once the subgraph is extracted, you will need to prepare it as the payload of different APIs for `invoke 
models for real-time inference
<https://docs.aws.amazon.com/sagemaker/latest/dg/realtime-endpoints-test-endpoints.html#realtime-endpoints-test-endpoints-api>`_.

> GraphStorm defines a specification of the payload contents.

Author: Good. Have a separated page for the specification.

.. _reat-time-payload-spec:

Payload content specification
******************************
The payload should be a JSON object in the format explained below. In the highest level, the JSON object
contains three fields: ``version``, ``gml_task``, and ``graph``.

``version`` (**Required**)
>>>>>>>>>>>>>>>>>>>>>>>>>>>
This field is used to identify the version of a specification, helping to avoid compatibility issues of different

> versions. The current version is ``gs-realtime-v0.1``.

Author: This version is about the Spec. Following the version convention, it is better to separate its version from the core code.

``gml_task`` (**Required**)
>>>>>>>>>>>>>>>>>>>>>>>>>>>
This field indicates what graph machine learning task this payload is for. Current specification supports two
options: 

* ``node_classification``
* ``node_regression``

``graph`` (**Required**)
>>>>>>>>>>>>>>>>>>>>>>>>>

This ``graph`` field is similar to fields, e.g., ``graph`` field, in :ref:`GSProcessing input specification
<gsprocessing_input_configuration>`. It contains three sub-fields, i.e., ``nodes``, ``edges``, and ``targets``.

A ``nodes`` field contains a list of ``node`` fileds. A ``node`` includes the raw input data values
of a node in the subgraph. It has the following required attributes.

* ``node_type``: string, the raw node type name in a graph. It should be same as these ``node_type`` defined in
  :ref:`gconstruct JSON specification <gconstruction-json>` or the ``type`` values of ``nodes`` defined in 
  :ref:`gsprocessing JSON specification <gsprocessing_input_configuration>`.
* ``node_id``: the raw node identifier.

> * ``features``: a dictionary, whose key is a feature name, and its value is the value of the feature.

Reviewer 2: ```suggestion
* ``features``: a dictionary, whose key is a feature name, and its value is the value of features.
```

>   feaure names should be same as these ``feature_name`` defined in :ref:`gconstruct JSON specification

Reviewer 2: ```suggestion
  feaure names should be same as the ``feature_name`` defined in :ref:`gconstruct JSON specification
```

  <gconstruction-json>`, or these ``name`` values of ``features`` fields defined in
  :ref:`gsprocessing JSON specification <gsprocessing_input_configuration>`.

An ``edges`` field contains a list of ``edge`` fields. An ``edge`` includes the raw input data values of an
edge in the subgraph. It has the following required attributes.

* ``edge_type``: list, the raw edge type name in the format of a list with three elements, which indicate

>   source node type, edge type, and destination edge type. It should be same as these ``relation`` fileds defined

Reviewer 2: ```suggestion
  source node type, edge type, and destination edge type. It should be same as the ``relation`` fileds defined
```

>   in :ref:`gconstruct JSON specification <gconstruction-json>` or these ``type`` values of ``source``

Reviewer 2: ```suggestion
  in :ref:`gconstruct JSON specification <gconstruction-json>` or the ``type`` values of ``source``
```

  ``relation``, and ``dest`` fileds defined in :ref:`gsprocessing JSON specification <gsprocessing_input_configuration>`.
* ``src_node_id``: user defined node identifier for the source node.
* ``dest_node_id``: user defined node identifier for the destination node.
* ``features``: a dictionary, whose key is a feature name, and its key is value of the feature. 
  feaure names should be same as these ``feature_name`` defined in :ref:`gconstruct JSON specification
  <gconstruction-json>`, or these ``name`` values of ``features`` fields defined in
  :ref:`gsprocessing JSON specification <gsprocessing_input_configuration>`.

A ``targets`` field contains a list of target ``node`` or ``edge`` fileds depending on the value of ``gml_task``

> These ``node`` or ``edge`` fileds is same as ``node`` and ``edge`` above, but the features field is not

Reviewer 2: Do not understand what you are talking about.

Author: Explain this in the new page and add an example to explain. The last sentence is want to make sure the target nodes or edges are in the graph's nodes or edges.

required. And they should be in the ``nodes`` or ``edges`` list of a ``graph``.

> An example payload JSON object is like the following:

Author: Will update this, and using `... inference payload ...`.

.. code:: yaml

    {
        "version": "gs-realtime-v0.1",
        "gml_task": "node_classification",
        "graph": {
            "nodes": [
                {
                    "node_type": "author",
                    "features": {
                        "feat": [

>                             0.011269339360296726,

Reviewer 2: Better to have a feat example of categorical features.

Author: added two categorical feature examples

The JSON object always include a ``message`` field, which provide additional information when the ``status_code`` is 200.

``error``
>>>>>>>>>>>>
The JSON object always include an ``error`` field, which provide detailed explanations when the ```status_code`` is not 200.

``data``
>>>>>>>>>
When the ``status_code`` is 200, the JSON object includes a populated ``data`` field. Otherwise, the data field is empty.

A ``200`` status response includes a JSON object containing inference results, with a single field called ``results``.
The values of ``results`` is a list that includes the inference values for all nodes specified in the payload's
``target`` field.

In addtion to the ``node_type`` and ``node_id`` fields, which match those in the payload ``target`` field, each result
in the list include a ``prediction`` field. This field contains the inference results for each node or edge. For
classification tasks, the value of ``prediction`` is a list of logits that can be used with classification method such
as `argmax`. For regression tasks, the value of ``prediction`` is a list with a single element, which represents the
regression result.

An example of a successful inference response:

> .. code:: yaml

Author: In rst file, the rendered results of yaml is more clear than results of json. So here choose yaml.

    # Decodes and prints the response body
    print(response['Body'].read().decode('utf-8'))

Endpoint invoke response
*************************

> As shown in the previous invoke example, the response from a GraphStorm real-time inference endpoint will include

Reviewer 2: ```suggestion
As shown in the previous example, the response from a GraphStorm real-time inference endpoint will include
```
